### DISCUSSION TITLE
Use layers for presentations?

### DISCUSSION THREAD

**Discussion by NelkonParker** [xournalpp/xournalpp#3602] [Q&A] [answered]
I want to use the layers functionality in Xournalpp to progressively build up a complete page for presentation purposes, but I can find no real documentation for it in the User Guide or FAQ.

For example, I have 4 layers:

Layer 1 says "Layers"
Layer 2 says "Layers are"
Layer 3 says "Layers are great"
Layer 4 says "Layers are great, aren't they?"

I progressively enable "show" on layers 2, 3, and 4 by pressing "Ctrl N" or similar, and the page gradually builds up to reveal the final message "Layers are great, aren't they?"

Is this functionality (or something equivalent) currently available in Xournalpp?

**Comment by rolandlo** [comment, marked as answer]
Yes, this functionality is available in Xournal++ 1.1.0. 
- First create the layers (I guess you know how to do it).
- Hide all layers except the first layer. There is a "LayerActions" plugin to do that automatically on all pages, which can be enabled in the plugin manger, but you can do it manually as well (e.g. in the sidebar).
- To uncover the next layer, you need to press `Shift + PageDown`. Do that until all layers are visible.

**Reply by NelkonParker** [reply]
Thanks. In fact, not 5 minutes ago, I discovered this for myself, and I was coming to add my own answer.

But I'm glad to have  it confirmed by someone that know what they're talking about.  This feature will help me a lot.

**Reply by NelkonParker** [reply]
>     * Hide all layers except the first layer. There is a "LayerActions" plugin to do that automatically on all pages, which can be enabled in the plugin manger, but you can do it manually as well (e.g. in the sidebar).

Could you expand a bit on how to use the plugin manager to do this? I know nothing about it.

**Reply by rolandlo** [reply]
In menu `Plugin > Plugin manager` activate the checkbox for the `LayerActions` plugin and restart Xournal++. Then in menu `Plugin` you should see an entry `Hide all layers except first layers and backgrounds`. Clicking on that menu entry does what it promises.

**Reply by NelkonParker** [reply]
Thanks. That's nice. However, it would be much nicer for me if there was a hotkey to enable that functionality, as I need to use it in full-screen mode. Is there some way to modify the plugin to bind it to a key sequence?

**Reply by rolandlo** [reply]
Sure, that can be modified in the `initUi` function of the plugin (that you find in `/usr/share/xournalpp/plugins/LayerActions/main.lua`)
Currently it reads:
```lua
function initUi()
  app.registerUi({["menu"] = "Clone non-background layers to next page", ["callback"] = "clone", ["accelerator"]="<Control><Shift>c"});
  app.registerUi({["menu"] = "Hide all layers except first layers and backgrounds", ["callback"] = "hide"});
  app.registerUi({["menu"] = "Add new top layer on each page", ["callback"] = "add"});
end
```
So cloning has a shortcut (accelerator) and hiding does not. Simply add an accelerator to the hiding, e.g. change it to

```lua
  app.registerUi({["menu"] = "Hide all layers except first layers and backgrounds", ["callback"] = "hide", ["accelerator"]="<Control><Shift>h"});
```

**Reply by NelkonParker** [reply]
Thanks. That's very useful. 

I'll have to take a look at plugins - they look like they could be very useful. Never used Lua though, and my software dev. days are long in the past.

**Reply by rolandlo** [reply]
There is some documentation on how to write plugins on the [website](https://xournalpp.github.io/guide/plugins/plugins/)
Lua is quite an easy language to get started with and the Xournal++ plugin API is limited, but relatively easy to use.
